### PR TITLE
The Newest Furry Race [Skeletons]

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR 
The skeleton race is a new race with a variety of unique game play and design factors surrounding them. They serve as a specialized high risk/reward setup that is weak enough to not be meta defining in any way.

Skeletons have incredible ecological durability, not needing pesky human things like food, water, air. They're also not susceptible to things like disease and cold temperatures. But for what they gain in environmental strength, they lose physically. 

Skeletons are incredibly physically weak, able to be "boned" in just a few hits with a heavy weapon. They also have no stomach, meaning that they cannot drink healing chemicals, forcing them to use alternate forms of healing (bible, bruise packs, medical bed)

As a helpful effect though, skeletons can never truly die. When a skeleton dies, they fall apart into their individual pieces, with the consciousness remaining in the skull (budget pai). Their body can then be reassembled if all the individual pieces are found. The only thing that's going to end a skeleton is a crematorium or a black hole. (or their own boredom, i suppose)

**To Do**
- [x] Finish up skeleton reassembly system

- [x] Window dressing and making systems not feel awful

**Screenshots**

**Changelog**

:cl:
- add: Added Skeletons, your lovely local undead spesspals. 97% less murderous than zombies and much easier to kill.

### CODE REVIEW

**Comment by EmoGarbage404** [comment]
this really isn't a ui change so the tag should honestly be removed.

**Comment by vulppine** [comment]
https://www.youtube.com/watch?v=ADLCBAKg2rY

**Comment by BlitzGunner225** [comment]
A race that takes away so many parts of the game is a bit iffy
I understand that it’s supped to be for people who want to just work nonstop so it kinda makes sense
I guess I’m for it

**Comment by Fishfish458** [comment]
Am I reading it right that onGib it actually saves a cloning entry to the cloning machine?

**Comment by freeman2651** [comment]
wouldn't particularly be a fan of this as a roundstart race especially (if what I am reading is right) that they are immune to the dangers of unprotected EVA. This would seem to imply I can space anything nearby to put anyone near me in trouble.
would be really funny as a specialty race though, I think that has a lot of potential

**Comment by EmoGarbage404** [comment]
They dont have barotrauma immunity. I think its either the same as humans or like slightly less.

**Comment by Fishfish458** [comment]
> They dont have barotrauma immunity. I think its either the same as humans or like slightly less.

If they were a non-starter special race, barotrauma proof skeletons would make a great boarding party.

**Comment by EmoGarbage404** [comment]
The food system is kinda secondary to the main mechanics of the race, so im going to open the pr for review and get back to that thing later.

File: Resources/Prototypes/species.yml
Comment (moonheart08) [inline Resources/Prototypes/species.yml:36]:
```suggestion
  roundStart: false #touch this and feel my wrath | bet --moony
```

Comment (EmoGarbage404) [reply Resources/Prototypes/species.yml:36]: Moony wasnt maintainer conduct talked about today during the meeting?
(Ps i am in your walls)

**Comment by EmoGarbage404** [comment]
I thought about my code and realized that its kind of awful and ungeneralized and desperately needs to be split into more components. Ill address that in an update soon.

File: Content.Server/Body/Components/BodyComponent.cs
Comment (mirrorcult) [inline Content.Server/Body/Components/BodyComponent.cs:86]: this is a perennial issue with body stuff (storing comps instead of entities) but these should return `HashSet<EntityUid>` rather than `List<SharedBodyPartComponent>`

File: Content.Server/Body/Components/SkeletonBodyManagerComponent.cs
Comment (mirrorcult) [inline Content.Server/Body/Components/SkeletonBodyManagerComponent.cs:26]: same here, store a list (or better yet hashset) of entities instead of body parts

File: Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs
Comment (mirrorcult) [inline Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs:38]: if this is intended to be used by the skeleton themselves (i assume it is) then this should really use an action rather than an alt verb.

actions for innate/self-verbs, verbs for things other people can do to you

File: Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs
Comment (mirrorcult) [inline Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs:93]: use hashsets for collections of entityuids

File: Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs
Comment (mirrorcult) [inline Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs:104]: i feel like these loops should be the other way around, iterate over nearby entities and check if any of them are parts you need. doesn't super matter since this doesnt exactly happen often anyway but

File: Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs
Comment (mirrorcult) [inline Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs:134]: use a dependency for `SharedHumanoidAppearanceSystem`

File: Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs
Comment (mirrorcult) [inline Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs:135]:
```suggestion
            MetaData(mob).EntityName = component.DNA.Value.Profile.Name;
```

File: Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs
Comment (mirrorcult) [inline Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs:27]: all entity systems have an innate reference to entity manager called `EntityManager`, you dont need this

File: Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs
Comment (mirrorcult) [inline Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs:153]:
```suggestion
        /// <param name="uid"> the entity the mind is going to be transfered which also stores the DNA</param>
```

File: Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs
Comment (mirrorcult) [inline Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs:154]:
```suggestion
        /// <param name="body">the entity whose DNA is being saved</param> 
```

File: Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs
Comment (mirrorcult) [inline Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs:168]: this is kind of really weird. but i cant think of a better way to do this

File: Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs
Comment (mirrorcult) [inline Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs:175]: dont store components on doafter events, just get it from the entityuid in the handler.

File: Content.Server/Body/Systems/TransferMindOnGibSystem.cs
Comment (mirrorcult) [inline Content.Server/Body/Systems/TransferMindOnGibSystem.cs:35]: invert control flow here

```suggestion
            if (!TryComp<MindComponent>(uid, out var mindComp) || mindComp?.Mind == null)
                return;

            if (!TryComp<BodyComponent>(uid, out var bodyComp))
                return;

            foreach (var part in bodyComp.Parts)
            {
                var entity = part.Key.Owner;
                if (HasComp<SkeletonBodyManagerComponent>(entity))
                {
                    component.TransferTarget = entity;
                    _skeletonBodyManager.UpdateDNAEntry(entity, uid);
                    mindComp.Mind.TransferTo(entity);
                }
            }
```

this is just generally a convention throughout the codebase since highly-intended control flow is really garbage to read

File: Content.Server/Body/Systems/TransferMindOnGibSystem.cs
Comment (mirrorcult) [inline Content.Server/Body/Systems/TransferMindOnGibSystem.cs:32]: these two components are highly coupled because of this and there's essentially no reason for them to be separate

File: Content.Shared/Body/Components/SharedBodyComponent.cs
Comment (mirrorcult) [inline Content.Shared/Body/Components/SharedBodyComponent.cs:397]: hashset

File: Content.Shared/Body/Part/BodyPartCompatibility.cs
Comment (mirrorcult) [inline Content.Shared/Body/Part/BodyPartCompatibility.cs:19]: why do you need this?

Comment (mirrorcult) [reply Content.Shared/Body/Part/BodyPartCompatibility.cs:19]: im not even sure why slimes were added to this tbh, thats not what this is for. these should just be biological

File: Content.Shared/Species/SpeciesPrototype.cs
Comment (mirrorcult) [inline Content.Shared/Species/SpeciesPrototype.cs:54]: i assume this is so you can have them all, like, be white, but still have different saturation/lightness? if so, then it should be docced

File: Resources/Prototypes/Body/Parts/skeleton.yml
Comment (mirrorcult) [inline Resources/Prototypes/Body/Parts/skeleton.yml:48]: the name for this component is kinda weird. again like i said it should be merged with the `TransferOnGib` and probably given a better name

Comment (mirrorcult) [reply Resources/Prototypes/Body/Parts/skeleton.yml:48]: also there's already an existing system where control gets sent to the brain when an entity is gibbed right? can that be used at all for this stuff instead. idr how it works tbh

File: Resources/Prototypes/Damage/modifier_sets.yml
Comment (mirrorcult) [inline Resources/Prototypes/Damage/modifier_sets.yml:126]: you missed `Piercing`. unless its supposed to just be 1.0

File: Resources/Prototypes/Damage/modifier_sets.yml
Comment (mirrorcult) [inline Resources/Prototypes/Damage/modifier_sets.yml:121]: this comment is incorrect

File: Resources/Prototypes/Entities/Mobs/Species/skeleton.yml
Comment (mirrorcult) [inline Resources/Prototypes/Entities/Mobs/Species/skeleton.yml:9]: are these really so different from humans that they cant inherit from them?

**Review by mirrorcult — CHANGES_REQUESTED** [review]

File: Resources/Prototypes/Entities/Mobs/Species/skeleton.yml
Comment (EmoGarbage404) [reply Resources/Prototypes/Entities/Mobs/Species/skeleton.yml:9]: The reason they can't inherit from humans is because they don't have hunger or thirst. That's honestly the only reason at this point.

File: Resources/Prototypes/Body/Parts/skeleton.yml
Comment (EmoGarbage404) [reply Resources/Prototypes/Body/Parts/skeleton.yml:48]: i looked into brain component and it seems to be kinda wacky. Not only does it not work for bodyparts, only for organs, it also allows the player to freely move from ghost into the brain, which is obviously broken as hell for a living character.

File: Resources/Prototypes/Entities/Mobs/Species/skeleton.yml
Comment (metalgearsloth) [reply Resources/Prototypes/Entities/Mobs/Species/skeleton.yml:9]: This will immediately lead to bugs whenever human gets updated and people don't realise skeletons also need updating so it needs fixing before a merge.

File: Resources/Prototypes/species.yml
Comment (metalgearsloth) [inline Resources/Prototypes/species.yml:36]: I don't feel these are RP friendly enough to be a roundstart and if any maintainers disagree I'm gonna ask why.

Comment (EmoGarbage404) [reply Resources/Prototypes/species.yml:36]: Yes i'm aware and i've seen the err in my ways.

Comment (EmoGarbage404) [reply Resources/Prototypes/species.yml:36]: i had a moment and i'm changing it back

File: Resources/Prototypes/Entities/Mobs/Species/skeleton.yml
Comment (EmoGarbage404) [reply Resources/Prototypes/Entities/Mobs/Species/skeleton.yml:9]: I would have the inheritance if it was possible, but there's simply an issue of no possible way of defining components that should not be inherited by the parent. Until such a system exists, realistically the best thing to do is just to stay on top of it and try and update whenever possible. Usually there aren't too many large prs for species so it shouldn't be that bad to manage until a decent solution exists.

**Comment by ike709** [comment]
Seems pretty meh as a roundstart race ngl. They just seem like worse [gehennites](https://github.com/tgstation/tgstation/pull/43777), which would be a much cooler race.

**Comment by Peptide90** [comment]
> Seems pretty meh as a roundstart race ngl. They just seem like worse [gehennites](https://github.com/tgstation/tgstation/pull/43777), which would be a much cooler race that got hampered by BYOND.

Should not be round start imo, as was discussed on discord. Event and by way of industrial accident only. Important crew being skeleton is just RIP round. Can’t make head or tails of the comments on it being round start so

**Comment by EmoGarbage404** [comment]
> Seems pretty meh as a roundstart race ngl. They just seem like worse [gehennites](https://github.com/tgstation/tgstation/pull/43777), which would be a much cooler race that got hampered by BYOND.

It is not a roundstart race. That has already been agreed on.

File: Resources/Prototypes/species.yml
Comment (metalgearsloth) [reply Resources/Prototypes/species.yml]: a

File: Resources/Prototypes/Entities/Mobs/Species/skeleton.yml
Comment (metalgearsloth) [reply Resources/Prototypes/Entities/Mobs/Species/skeleton.yml:9]: I'll just cherry-pick from my kepori branch...

**Comment by metalgearsloth** [comment]
I got a bone to pick with this pr

**Comment by EmoGarbage404** [comment]
as per sloth's request, the licenses for bone_rattle.ogg and skeleton_scream.ogg have been added. PR should be good to be merged.

File: Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs
Comment (metalgearsloth) [reply Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs]: I just made it check for distance for each prat given we don't care what else is nearby.

**Pull request merged**
